internal/servers/controller/handlers/credentialstores/vault.go

```diff
@@ -0,0 +1,70 @@
+package credentialstores
+
+import (
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+// extractClientCertAndPk takes the values passed into the api for client
+// certificate and client certificate key and parses them into pem blocks.
+// Any non certificate pem blocks are treated as a private key. An error is
+// returned if there is more than 1 private key provided across both fields.
+func extractClientCertAndPk(cert, pk string) ([]*pem.Block, *pem.Block, error) {
+	const op = "credentialstores.extractClientCertAndPk"
+	pks, err := decodePemBlocks(pk)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op, errors.WithMsg("failed to parse client certificate private key"))
+	}
+	var pkPem *pem.Block
+	switch len(pks) {
+	case 0:
+	case 1:
+		pkPem = pks[0]
+	default:
+		return nil, nil, errors.New(errors.InvalidParameter, op, "private key payload contained multiple pem blocks")
+	}
+
+	bs, err := decodePemBlocks(cert)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op, errors.WithMsg("failed to parse client certificate into pem blocks"))
+	}
+	pkIdx := -1
+	for i, b := range bs {
+		if !strings.Contains(strings.ToLower(b.Type), "certificate") {
```

Comment:
FWIW, the PEM "standard" uses upper case for block declarations so if something is not uppercase it's likely to be unaccepted in a lot of implementations. Even Go's PEM decode function requires `BEGIN` and `END`, case-sensitively, for the block declaration. So technically you're allowing `-----BEGIN certificate-----` although it's unlikely such a block would be very portable.

I don't think this _needs_ to be changed, but I think doing a case sensitive comparison with no ToLower/ToUpper here is perfectly acceptable as well.

Reply:
@talanknight